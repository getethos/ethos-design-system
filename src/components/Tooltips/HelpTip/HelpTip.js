import React, { useState } from 'react'
import PropTypes from 'prop-types'
import styles from './HelpTip.module.scss'

export const HelpTip = ({
  labelCopy,
  helpCopy,
  disabled,
  position,
  alignLeftTo,
  ...rest,
}) => {

  const [hover, setHover ] = useState(false)

  // TODO --- maybe put this in use layout effect and only run it once so we don't repaint
  const getAlignLeft = () => {
    console.log('getAlignLeft called...')
    let styles = {}
    const el = document.querySelector(alignLeftTo)
    if (el) {
      const rect = el.getBoundingClientRect();
      const win = el.ownerDocument.defaultView;
      const leftCalculated = rect.left + win.pageXOffset
      styles.left = `${leftCalculated}px`
      styles.width = `calc(100vw - ${leftCalculated + 96}px)`

      if (rect.top < 72) {
        styles.top= `${rect.top + 96}px`
      } else {
        styles.top= `${rect.top - 16}px` //Note--we're using position fixed so no need to add pageYOffSet
      }
      styles.position = 'fixed'
      styles.transform = 'translate(0%, -100%)'
      styles.overflow = 'visible'
      styles.zIndex = '1'
    }
    return styles
  }

  const toggleHover = () => {
    console.log('toggleHover called...')
    setHover(!hover)
  }

  const getDynamicStyles = () => {
    let styles = {
      // minWidth: rest['min-width'],
    }
    const alignLeftEl = alignLeftTo
    if (alignLeftTo) {
      styles = Object.assign({}, styles, getAlignLeft())
    }
    return styles
  }

  return (
    <div aria-label={labelCopy} data-tid={rest['data-tid']} onMouseEnter={toggleHover} onMouseLeave={toggleHover}>
      <span className={styles.info} data-pos={position}>
        i
      </span>
      <span className={styles.tip} style={ getDynamicStyles() } >
        {helpCopy}
      </span>
    </div>
  )
}

HelpTip.propTypes = {
  className: PropTypes.string,
  labelCopy: PropTypes.string.isRequired,
  helpCopy: PropTypes.string.isRequired,
  'data-tid': PropTypes.string.isRequired,
  'min-width': PropTypes.number,
  disabled: PropTypes.bool,
  position: PropTypes.string,
  alignLeftTo: PropTypes.string,
}

HelpTip.defaultProps = {
  position: 'above',
  // 'min-width': 500,
}
