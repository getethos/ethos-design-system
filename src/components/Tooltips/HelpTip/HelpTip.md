HelpTip

Above

```jsx
import styles from './example.module.scss'
;<div className={styles.example}>
  <HelpTip
    labelCopy="Aria label"
    helpCopy="Including skin cancer (including basal cell and squamous cell), melanoma, lymphoma, leukemia, tumor, polyp, any other abnormal growth, any other type of cancer"
    data-tid="the-help-tip"
    position="above"
  />
</div>
```

Belowâ€”`min-width` defaults to 500, but you can override that.

```jsx
import styles from './example.module.scss'
;<div className={styles.example}>
  <HelpTip
    min-width={800}
    className="parent"
    labelCopy="aria label"
    helpCopy="Including skin cancer (including basal cell and squamous cell), melanoma, lymphoma, leukemia, tumor, polyp, any other abnormal growth, any other type of cancer"
    data-tid="the-help-tip"
    position="below"
  />
</div>
```

Left

```jsx
import styles from './example.module.scss'
;<div className={styles.example}>
  <HelpTip
    labelCopy="aria label"
    helpCopy="Including skin cancer (including basal cell and squamous cell), melanoma, lymphoma, leukemia, tumor, polyp, any other abnormal growth, any other type of cancer"
    data-tid="the-help-tip"
    position="left"
  />
</div>
```

Right

```jsx
import styles from './example.module.scss'
;<div className={styles.example}>
  <HelpTip
    labelCopy="aria label"
    helpCopy="Including skin cancer (including basal cell and squamous cell), melanoma, lymphoma, leukemia, tumor, polyp, any other abnormal growth, any other type of cancer"
    data-tid="the-help-tip"
    position="right"
  />
</div>
```
