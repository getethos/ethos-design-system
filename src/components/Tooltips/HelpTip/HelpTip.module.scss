/* TODO ONCE Css modules land */
/* @import "../Colors.scss"; */
/* @import '../Spacer/Spacer.scss'; */

[aria-label] {
  position: relative;
  cursor: pointer;
}

.info {
  top: -2px;
  padding: var(--Space-4) 0px 0px;
  background: var(--GraySecondary--translucent);
  color: white;
  border-radius: 50%;
  font-size: 14px;
  font-weight: 400;
  width: 18px;
  height: 18px;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  user-select: none;
}

.tip {
  position: absolute;
  top: -5px;
  left: 0px;
  pointer-events: none;
  background-color: white;
  border: 1px solid rgb(230, 230, 230);
  padding: 5px;
  will-change: transform;
  transform: translate3d(0px, -44px, 0px);
  opacity: 0;
}

[aria-label]:hover [data-pos="above"] + span {
  opacity: 1;
  /* TODO: Maybe we make all these smarter w/calc or something?? */
  transform: translate(-50%, -100%);
}

[aria-label]:hover [data-pos="below"] + span {
  opacity: 1;
  transform: translate(-50%, 65%);
}

[aria-label]:hover [data-pos="left"] + span {
  opacity: 1;
  transform: translate(-101%, -25%);
}

[aria-label]:hover [data-pos="right"] + span {
  opacity: 1;
  transform: translate(5%, -25%);
}
