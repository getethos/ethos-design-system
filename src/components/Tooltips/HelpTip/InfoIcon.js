/* eslint-disable  */
import React from 'react'
import PropTypes from 'prop-types'

// More code from monorepo

function isTouchUI() {
  return 'ontouchstart' in window
}

function InfoIcon({ onMouseOver, onMouseOut, onTouch, onFocus, onBlur }) {
  return (
    <div
      style={{
        position: 'relative',
        top: '-2px',
        padding: isTouchUI() ? '2px 8px 8px 8px' : '0',
        marginLeft: isTouchUI() ? '0' : '5px',
        display: 'inline-flex',
        alignItems: 'center',
        justifyContent: 'center',
      }}
      onMouseOver={onMouseOver}
      onMouseOut={onMouseOut}
      onTouchStart={onTouch}
      onFocus={onFocus}
      onBlur={onBlur}
      onClick={
        (e) => e.preventDefault() /* Do not trigger click event on parent */
      }
    >
      <div
        style={{
          top: '-2px',
          padding: '4px 0 0 0',
          background: 'rgba(0, 0, 0, 0.78)',
          color: 'white',
          borderRadius: '50%',
          fontSize: '14px',
          fontWeight: '800',
          width: '18px',
          height: '18px',
          display: 'inline-flex',
          alignItems: 'center',
          justifyContent: 'center',

          // Prevent touch users from selecting the "i" text when held
          WebkitTouchCallout: 'none',
          WebkitUserSelect: 'none',
          khtmlUserSelect: 'none',
          MozUserSelect: 'none',
          msUserSelect: 'none',
          userSelect: 'none',
        }}
      >
        i
      </div>
    </div>
  )
}

InfoIcon.defaultProps = {
  onMouseOver: () => {},
  onMouseOut: () => {},
  onTouchStart: () => {},
  onTouchEnd: () => {},
  onFocus: () => {},
  onBlur: () => {},
}

InfoIcon.propTypes = {
  onMouseOver: PropTypes.func,
  onMouseOut: PropTypes.func,
  onTouchStart: PropTypes.func,
  onTouchEnd: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
}

export default InfoIcon
