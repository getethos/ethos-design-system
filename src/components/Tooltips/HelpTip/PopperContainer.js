import React from 'react'
import PropTypes from 'prop-types'
import { Manager, Reference, Popper } from 'react-popper'
import debounce from './LodashDebounce'

// From monorepo

export class PopperContainer extends React.Component {
  static propTypes = {
    tooltipText: PropTypes.string.isRequired,
    children: PropTypes.func.isRequired,
  }

  constructor(props) {
    super(props)
    this.togglePopup = this.togglePopup.bind(this)
  }

  state = {
    showPopup: false,
  }

  openPopup = () => {
    this.setState({
      showPopup: true,
    })
  }

  closePopup = () => {
    this.setState({
      showPopup: false,
    })
  }

  togglePopup = debounce(
    () => {
      this.setState({
        showPopup: !this.state.showPopup,
      })
    },
    500,
    { leading: true, trailing: false }
  )

  render() {
    const { showPopup } = this.state
    const { children, tooltipText } = this.props

    return (
      <Manager>
        <div style={{ position: 'relative', zIndex: showPopup ? '1' : '0' }}>
          <Reference>
            {({ ref }) => (
              <div ref={ref}>
                <Popper placement="top">
                  {({ ref: innerRef, style, placement, ...rest }) => {
                    const finalStyle = {
                      ...style,
                      backgroundColor: 'white',
                      border: '1px solid rgb(230, 230, 230)',
                      padding: '5px',
                      pointerEvents: 'none',
                      top: '-5px',
                    }
                    if (!showPopup) finalStyle.visibility = 'hidden'
                    return (
                      <div style={finalStyle} ref={innerRef}>
                        {tooltipText}
                      </div>
                    )
                  }}
                </Popper>
                {children({
                  openPopup: this.openPopup,
                  closePopup: this.closePopup,
                  togglePopup: this.togglePopup,
                })}
              </div>
            )}
          </Reference>
        </div>
      </Manager>
    )
  }
}
