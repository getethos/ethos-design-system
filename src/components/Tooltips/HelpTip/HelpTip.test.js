import React from 'react'
import { HelpTip } from './HelpTip'
import renderer from 'react-test-renderer'

const formChangeHandlerStub = jest.fn()

describe('HelpTip', () => {
  it('default rendering', () => {
    const tree = renderer
      .create(
        <HelpTip
          labelCopy="Aria label copy"
          helpCopy="this is the tip message"
          data-tid="le-help-tip"
          position="above"
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
