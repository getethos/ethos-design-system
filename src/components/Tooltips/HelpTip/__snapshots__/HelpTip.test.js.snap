// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelpTip default rendering 1`] = `
<div
  aria-label="Aria label copy"
  data-tid="le-help-tip"
>
  <span
    className="info"
    data-pos="above"
  >
    i
  </span>
  <span
    className="tip"
    style={
      Object {
        "minWidth": 500,
      }
    }
  >
    this is the tip message
  </span>
</div>
`;
